fxrunner/src/lib/splash.rs

```diff
@@ -0,0 +1,191 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::ffi::CStr;
+use std::io;
+use std::ptr::null_mut;
+use std::thread;
+
+use async_trait::async_trait;
+use lazy_static::lazy_static;
+use tokio::sync::oneshot;
+use winapi::shared::minwindef::DWORD;
+use winapi::shared::windef::HWND;
+use winapi::um::winuser::{MSG, WNDCLASSA};
+use winapi::um::{processthreadsapi, wingdi, winuser};
+
+use crate::osapi::error::{check_nonnull, check_nonzero};
+use crate::osapi::handle::get_module_handle;
+
+lazy_static! {
+    static ref WINDOW_CLASS_NAME: &'static CStr =
+        CStr::from_bytes_with_nul(&b"fxrunnerbg\0"[..]).unwrap();
+}
+
+#[async_trait]
+pub trait Splash: Sized {
+    async fn new(display_widht: u32, display_height: u32) -> Result<Self, io::Error>;
+    fn destroy(&mut self) -> Result<(), io::Error>;
+}
+
+/// A splash screen that covers the entire display.
+///
+/// The splash screen is painted a solid red (#FF0000) so that the Firefox Window
+/// can be easily differentiated from the background.
+pub struct WindowsSplash {
+    /// The thread ID of the UI thread, so that we may send messages to it via
+    /// `PostThreadMessageA` API.
+    ui_thread_id: DWORD,
+
+    /// The join handle for the thread.
+    ui_thread_join_handle: Option<thread::JoinHandle<()>>,
+}
+
+impl WindowsSplash {
+    /// Register the window class that `Splash` will use.
+    ///
+    /// Must be called before [`Splash::new()`][Splash::new] or it will fail.
+    ///
+    /// [Splash::new]: struct.Splash.html#method.new
+    pub fn register_window_class() -> Result<(), io::Error> {
```

Comment:
Can you use `GetClassInfo` to check if the window class is already registered, and then just always call this from `create_and_show_window`? That would allow you to not couple calling code to the internals here.

Reply:
I'll give that a shot